File: mobile/src/utils/sessionAcademique.ts
/**
 * Session académique — calcul par pays.
 *
 * Règle par défaut (Afrique francophone + anglophone de la région) :
 *   mois 1-5  (janv-mai) → année_précédente / année_courante   (ex : juillet 2026 → "2025-2026")
 *   mois 6-12 (juin-déc) → année_courante / année_suivante     (ex : septembre 2026 → "2026-2027")
 *
 * Variations futures possibles : certaines années scolaires démarrent en janvier
 * (hémisphère sud), etc. Structure extensible.
 */

import type { PaysCode } from '../data/schoolSystems';

type RentreeBreakpoint = {
  /** Mois (1-12) — à partir duquel la session "suivante" démarre */
  startMonth: number;
};

const RENTREE_PAR_PAYS: Partial<Record<PaysCode, RentreeBreakpoint>> = {
  CM: { startMonth: 6 }, // septembre mais on anticipe à juin (préinscriptions)
  CI: { startMonth: 6 },
  SN: { startMonth: 6 },
  GA: { startMonth: 6 },
  CG: { startMonth: 6 },
  CD: { startMonth: 6 },
  BJ: { startMonth: 6 },
  TG: { startMonth: 6 },
  BF: { startMonth: 6 },
  ML: { startMonth: 6 },
  NE: { startMonth: 6 },
  NG: { startMonth: 6 }, // rentrée septembre
  GH: { startMonth: 6 }, // rentrée septembre/janvier selon terms
};

const DEFAULT: RentreeBreakpoint = { startMonth: 6 };

export interface SessionInfo {
  annee: string;       // "2025-2026"
  debut: number;       // 2025
  fin: number;         // 2026
  pays: PaysCode;
}

export function computeSessionAcademique(pays: PaysCode = 'CM', at: Date = new Date()): SessionInfo {
  const cfg = RENTREE_PAR_PAYS[pays] ?? DEFAULT;
  const year = at.getFullYear();
  const month = at.getMonth() + 1; // 1-12

  if (month >= cfg.startMonth) {
    return { annee: `${year}-${year + 1}`, debut: year, fin: year + 1, pays };
  }
  return { annee: `${year - 1}-${year}`, debut: year - 1, fin: year, pays };
}

/** Parse "2025-2026" → { debut:2025, fin:2026 } ; null si invalide */
export function parseSessionString(str: string): { debut: number; fin: number } | null {
  const m = str.match(/^(\d{4})\s*[-–]\s*(\d{4})$/);
  if (!m) return null;
  const debut = parseInt(m[1], 10);
  const fin = parseInt(m[2], 10);
  if (fin !== debut + 1) return null;
  return { debut, fin };
}

/** true si la session détectée sur le document correspond à celle calculée */
export function isSessionCoherente(detected: string, expected: SessionInfo): boolean {
  const p = parseSessionString(detected);
  if (!p) return false;
  return p.debut === expected.debut && p.fin === expected.fin;
}

File: backend/src/utils/session_academique.rs
//! Session académique — calcul par pays.
//!
//! Règle par défaut (Afrique francophone + anglophone de la région) :
//!   mois 1-5  (janv-mai) → année_précédente / année_courante
//!   mois 6-12 (juin-déc) → année_courante / année_suivante
//!
//! Miroir de `frontend/src/utils/sessionAcademique.ts`.

use chrono::{Datelike, Utc};

/// Mois de bascule (1-12) par code pays ISO-2.
fn rentree_start_month(pays: &str) -> u32 {
    match pays {
        "CM" | "CI" | "SN" | "GA" | "CG" | "CD" | "BJ" | "TG" | "BF" | "ML" | "NE" | "NG"
        | "GH" => 6,
        _ => 6,
    }
}

#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub annee: String, // "2025-2026"
    pub debut: i32,
    pub fin: i32,
    pub pays: String,
}

/// Calcule la session académique pour un pays à une date donnée.
pub fn compute_session_academique(pays: &str, at: Option<chrono::DateTime<Utc>>) -> SessionInfo {
    let at = at.unwrap_or_else(Utc::now);
    let year = at.year();
    let month = at.month();
    let start = rentree_start_month(pays);
    let pays = pays.to_string();

    if month >= start {
        SessionInfo {
            annee: format!("{}-{}", year, year + 1),
            debut: year,
            fin: year + 1,
            pays,
        }
    } else {
        SessionInfo {
            annee: format!("{}-{}", year - 1, year),
            debut: year - 1,
            fin: year,
            pays,
        }
    }
}

/// Parse "2025-2026" → (2025, 2026), None si invalide.
pub fn parse_session_string(s: &str) -> Option<(i32, i32)> {
    let parts: Vec<&str> = s.split(|c| c == '-' || c == '–').map(str::trim).collect();
    if parts.len() != 2 {
        return None;
    }
    let debut: i32 = parts[0].parse().ok()?;
    let fin: i32 = parts[1].parse().ok()?;
    if fin != debut + 1 {
        return None;
    }
    Some((debut, fin))
}

pub fn is_session_coherente(detected: &str, expected: &SessionInfo) -> bool {
    match parse_session_string(detected) {
        Some((d, f)) => d == expected.debut && f == expected.fin,
        None => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    #[test]
    fn september_2026_cm() {
        let s = compute_session_academique(
            "CM",
            Some(Utc.with_ymd_and_hms(2026, 9, 1, 0, 0, 0).unwrap()),
        );
        assert_eq!(s.annee, "2026-2027");
    }

    #[test]
    fn march_2026_cm() {
        let s = compute_session_academique(
            "CM",
            Some(Utc.with_ymd_and_hms(2026, 3, 1, 0, 0, 0).unwrap()),
        );
        assert_eq!(s.annee, "2025-2026");
    }

    #[test]
    fn parse_ok() {
        assert_eq!(parse_session_string("2025-2026"), Some((2025, 2026)));
        assert_eq!(parse_session_string("2025 – 2026"), Some((2025, 2026)));
        assert_eq!(parse_session_string("2025-2027"), None);
    }
}
